### PR TITLE
AWS-S3 integration

### PR DESCRIPTION
**Pull request by itaimajlin**
Related to https://github.com/demisto/etc/issues/10325

documentation: https://github.com/demisto/etc/issues/11045
need to add docker container with boto3 python lib

### CODE REVIEW

File: Integrations/integration-AWS-S3.yml [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        ec = {'AWS.S3.NewBucket': data}
+        return create_entry('Created S3 Bucket', data, ec)
+
+    def delete_bucket(args):
+        data =[]
+        response = s3.delete_bucket(Bucket=args.get('bucketName'))
+        data.append ({
+            'BucketName': bucket
+            })
+        ec = {'AWS.S3.DeletedBuckets': data}
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: I don't think that there is a need for `DeletedBuckets` context, but I think we should delete that bucket from `Bucket` list if exists there

File: Integrations/integration-AWS-S3.yml [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                        Bucket=args.get('bucketName'),
+                        CreateBucketConfiguration={
+                            'LocationConstraint': args.get('awsRegion')
+                        })
+        data.append ({
+            'BucketName': args.get('bucketName'),
+            'AWSRegion': args.get('awsRegion')
+        })
+
+        ec = {'AWS.S3.NewBucket': data}
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: Just add new bucket to `Bucket` context - no need for `NewBucket`

File: Integrations/integration-AWS-S3.yml [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+      description: The name of the AWS S3 bucket
+    description: List all S3 buckets in AWS account
+  - name: aws-s3-get-bucket-policy
+    arguments:
+    - name: bucketName
+      required: true
+      description: Name of bucket
+      isArray: true
+    outputs:
+    - contextPath: AWS.S3.Policy.Version
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: I think maybe the policy object should be contained in Bucket object, like `AWS.S3.Bucket.Policy....`

File: Integrations/integration-AWS-S3.yml [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+      isArray: true
+    outputs:
+    - contextPath: AWS.S3.DeletedBuckets.BucketName
+      description: Name of S3 bucket that was deleted
+      type: string
+    description: Delete AWS S3 bucket
+  - name: aws-s3-list-buckets
+    arguments: []
+    outputs:
+    - contextPath: AWS.S3.BucketList.BucketName
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: `BucketList` => `Bucket`

File: Integrations/integration-AWS-S3.yml [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+
+    def set_bucket_policy(args):
+        bucket_policy = json.dumps(args.get('bucketPolicyJson'))
+        s3.put_bucket_policy(Bucket=args.get('bucketName'), Policy=bucket_policy)
+        return 'Successfully applied Bucket policy to %s bucket' % args.get('BucketName')
+
+    def delete_bucket_policy(args):
+        s3.delete_bucket_policy(Bucket=args.get('bucketName'))
+        return 'Policy deleted from %s' % args.get('bucketName')
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: should delete the policy from `Policy` context if such object exists in the context

File: Integrations/integration-AWS-S3.yml [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+                'Resource': statement['Resource'],
+                'Effect': statement['Effect']
+                })
+
+        ec = {'AWS.S3.Policy': data, 'AWS.S3.Policy.json': response['Policy']}
+        return create_entry('AWS S3 Bucket Policy', data, ec)
+
+
+    def set_bucket_policy(args):
+        bucket_policy = json.dumps(args.get('bucketPolicyJson'))
```
Comment (anara123) [inline Integrations/integration-AWS-S3.yml]: check if `bucketPolicyJson` is dict or string. If dict then dump to json, if string then validate json validity and then pass  to `s3.put_bucket_policy`

Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

File: Integrations/integration-AWS-S3.yml (hunk 5 above)
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: will need help with that, how do you delete an object from context?

File: Integrations/integration-AWS-S3.yml (hunk 4 above)
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

File: Integrations/integration-AWS-S3.yml (hunk 3 above)
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

File: Integrations/integration-AWS-S3.yml (hunk 2 above)
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

File: Integrations/integration-AWS-S3.yml (hunk 1 above)
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: gonna need help with deleting an object from context

File: Integrations/integration-AWS-S3.yml [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+            })
+        ec = {'AWS.S3.DeletedBuckets': data}
+        return create_entry('The following S3 bucket was deleted', data, ec)
+
+    def list_buckets(args):
+        data = []
+        response = s3.list_buckets()
+        for bucket in response['Buckets']:
+            data.append({
+                'BucketName': bucket['Name']
```
[the remark this answers is not in the document]
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: the bucket variable is initialized in the for loop (for bucket in...)

File: Integrations/integration-AWS-S3.yml [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        })
+
+        ec = {'AWS.S3.NewBucket': data}
+        return create_entry('Created S3 Bucket', data, ec)
+
+    def delete_bucket(args):
+        data =[]
+        response = s3.delete_bucket(Bucket=args.get('bucketName'))
+        data.append ({
+            'BucketName': bucket
```
[the remark this answers is not in the document]
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

File: Integrations/integration-AWS-S3.yml [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+      - public-read-write
+      - authenticated-read
+      description: ACL for S3 bucket
+      defaultValue: private
+    - name: awsRegion
+      required: true
+      description: AWS Region to create bucket (optional, e.g us-west-1) if no region
+        is selected the default region will be used
+    outputs:
+    - contextPath: AWS.S3.NewBucket.BucketName
```
[the remark this answers is not in the document]
Comment (itaimajlin) [reply Integrations/integration-AWS-S3.yml]: fixed

**Comment by itaimajlin** [comment]
i will add a test playbook but it shouldn't run automatically since it will create AWS resources which cost money

**Comment by itaimajlin** [comment]
Add test platybook and pushed docker image to docker hub (demist/boto3)
user for tests:
https://keepersecurity.com/vault/#detail/4-XG2y2PtrkJsqH0yz27Xg

**Pull request merged**